device/esp_tinyusb/include/tinyusb.h

```diff
@@ -31,7 +31,20 @@ typedef struct {
     const char **string_descriptor;            /*!< Pointer to array of string descriptors. If set to NULL, TinyUSB device will use a default string descriptors whose values are set in Kconfig */
     int string_descriptor_count;               /*!< Number of descriptors in above array */
     bool external_phy;                         /*!< Should USB use an external PHY */
-    const uint8_t *configuration_descriptor;   /*!< Pointer to a configuration descriptor. If set to NULL, TinyUSB device will use a default configuration descriptor whose values are set in Kconfig */
+    union {
+        struct {
+            const uint8_t *configuration_descriptor; /*!< Pointer to a configuration descriptor. If set to NULL, TinyUSB device will use a default configuration descriptor whose values are set in Kconfig */
+            const uint8_t *reserved1;                /*!< For backward compatibility */
+            const uint8_t *reserved2;                /*!< For backward compatibility */
```

Comment:
I can still see some `reserved` fields. Why are they there?

Reply:
Briefly: to exclude the undefined behavior.

In general, the idea to make the union of two structs with one and two pointers means, that during the configuration and structure initialization (`const tinyusb_config_t tusb_cfg = { ... }`)  we use old `.configuration_descriptor` member and we can use new member for `.fs_cfg_desc`. 

Problem appears when compiler creates the structure and fill the memory. To be more specific - in the way how he solves the two-structs union initialization. If compiler makes it one after another, that means, that is we are initialize not all members of structure, then the memory will be initialized with the last one. 

Here you may find the simple example-code (to verify it in OnlineGDB, for example): 
```
static uint8_t const config_desc[] = {
    'a', 'b', 'c',
};

static uint8_t const fs_config_desc[] = {
    'w', 's', 'x'
};

typedef struct {
    union { 
        struct {
            const uint8_t *ptr1_1; 
            const uint8_t *ptr1_2;
        };
        struct {
            const uint8_t *ptr2_1;   
            const uint8_t *ptr2_2;              
        };
    };
} descriptor_t; 

void print_desc(const uint8_t * ptr) 
{
    if (ptr == NULL) {
        printf("NULL\r\n");
        return;
    }
    for (int i = 0; i < 3; i++) {
        printf("%c ", ptr[i]);
    }
    printf("\n\r");
}

int main()
{
    const descriptor_t d = {
        .ptr1_1 = config_desc,
        // .ptr1_2 = config_desc,
        // .ptr2_1 = fs_config_desc,
        .ptr2_2 = fs_config_desc
    };
    
    printf("size %ld\r\n", sizeof(descriptor_t));
    
    for(int i = 0; i < sizeof(descriptor_t); i++) {
        printf("%02X ", *((uint8_t*)(&d) + i));
    }
    printf("\r\n");
    
    print_desc(d.ptr1_1);
    print_desc(d.ptr1_2);
    print_desc(d.ptr2_1);
    print_desc(d.ptr2_2);

    return 0;
}
```

The result of the code: 
```
size 16
00 00 00 00 00 00 00 00 0A 40 A4 CB 52 56 00 00 
NULL
w s x 
NULL
w s x 
```

We can see, that the structure contains the data of the initialization values for second structure in union (`.ptr2_2 = fs_config_desc` and `.ptr2_1` is not defined ).  

For RISCV it works a bit differentle and when we have the `.reserved` field, the initialization of `.fs_cfg_desc` doesn't rewrite the `.configuration_descriptor` member with `.hs_cfg_desc` (which is not defined, as we don't change the example code).

So, either we need to force to change the example and use always new name for `.hs_cfg_desc`, or add `.reserved` field. 
I decided to do the second option, so we can postpone with updating the example code.

Reply:
@roma-jam you **cannot initialize multiple members of a union!**

Unfortunately, GCC will not warn you about this. It is the programmer's responsibility. Clang based compilers will give you a warning in C:
```
warning: initializer overrides prior initialization of this subobject [-Winitializer-overrides]
<source>:41:9: note: previous initialization is here
        .ptr1_1 = config_desc,
        ^~~~~~~~~~~~~~~~~~~~~
```

Just for completeness, you won't be able to compile the same code with C++ compilers, which are more strict.
Most of them (GCC based) will get you this error:
```
error: too many initializers for 'const descriptor_t'
```
Which is correct because you **cannot initialize multiple members of a union!**
Some other compilers (Clang based) give a better error message:
```
error: initializer overrides prior initialization of this subobject [-Winitializer-overrides]
<source>:40:9: note: previous initialization is here
        .ptr1_1 = config_desc,|
```

Reply:
Thanks for the detailed explanation!